sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,388 @@
+        int chunk_idx, n_samples_chunk_eff
+        int start, end
+        # int num_threads
+
+        int j, k
+
+    # If n_samples < 256 there's still one chunk of size n_samples_r
+    if n_chunks == 0:
+        n_chunks = 1
+        n_samples_chunk = 0
+
+    # re-initialize all arrays at each iteration
+    centers_squared_norms = row_norms(centers_new, squared=True)
+
+    if update_centers:
+        memcpy(&centers_old[0, 0], &centers_new[0, 0], n_clusters * n_features * sizeof(floating))
+        memset(&centers_new[0, 0], 0, n_clusters * n_features * sizeof(floating))
+        memset(&weight_in_clusters[0], 0, n_clusters * sizeof(floating))
+
+    # set number of threads to be used by openmp
+    # num_threads = n_jobs if n_jobs != -1 else openmp.omp_get_max_threads()
+
+    for chunk_idx in prange(n_chunks, nogil=True, num_threads=n_jobs):
+        # remaining samples added to last chunk
+        if chunk_idx == n_chunks - 1:
+            n_samples_chunk_eff = n_samples_chunk + n_samples_r
+        else:
+            n_samples_chunk_eff = n_samples_chunk
+
+        start = chunk_idx * n_samples_chunk
+        end = start + n_samples_chunk_eff
+
+        _update_chunk_dense(
+            &X[start, 0],
+            sample_weight[start: end],
+            x_squared_norms[start: end],
+            centers_old,
+            centers_new,
+            centers_squared_norms,
+            weight_in_clusters,
+            labels[start: end],
+            update_centers)
+
+    if update_centers:
+        _relocate_empty_clusters_dense(X, sample_weight, centers_old,
+                                       centers_new, weight_in_clusters, labels)
+
+        _average_centers(centers_new, weight_in_clusters)
+        _center_shift(centers_old, centers_new, center_shift)
+
+
+cdef void _update_chunk_dense(floating *X,
+                              floating[::1] sample_weight,
+                              floating[::1] x_squared_norms,
+                              floating[:, ::1] centers_old,
+                              floating[:, ::1] centers_new,
+                              floating[::1] centers_squared_norms,
+                              floating[::1] weight_in_clusters,
+                              int[::1] labels,
+                              bint update_centers) nogil:
+    """K-means combined EM step for one dense data chunk.
+
+    Compute the partial contribution of a single data chunk to the labels and
+    centers.
+    """
+    cdef:
+        int n_samples = labels.shape[0]
+        int n_clusters = centers_old.shape[0]
+        int n_features = centers_old.shape[1]
+
+        floating sq_dist, min_sq_dist
+        int i, j, k, label
+
+        floating *pairwise_distances_ptr = <floating*> malloc(n_samples * n_clusters * sizeof(floating))
+        floating[:, ::1] pairwise_distances
+
+    with gil:
+        pairwise_distances = <floating[:n_samples, :n_clusters:1]> pairwise_distances_ptr
```

Comment:
That's a cool casting! Questions:

- Are you using it just because it makes the code easier to read (very valid reason) or it also has some implicit impact?
- Any reason you're not using this as well for the X pointer?
- Have you noticed any difference in performance / Python interaction (when annotating the code)?

Reply:
> Are you using it just because it makes the code easier to read (very valid reason) or it also has some implicit impact?

For code readability. I didn't notice any drop of performance.

> Any reason you're not using this as well for the X pointer?

I should try, but I'm afraid that if X is read only, taking a fused type view will break somehow...